src/Compilers/Test/Utilities/CSharp/CompilationTestUtils.cs

```diff
@@ -278,5 +278,76 @@ public static List<string> LookupNames(this SemanticModel model, int position, I
                     : model.LookupSymbols(position, container);
             return symbols.Select(s => s.Name).Distinct().ToList();
         }
+
+        internal static void VerifyTypes(this CSharpCompilation compilation)
+        {
+            foreach (var tree in compilation.SyntaxTrees)
+            {
+                VerifyTypes(compilation, tree);
+            }
+        }
+
+        internal static void VerifyTypes(this CSharpCompilation compilation, SyntaxTree tree)
+        {
+            var root = tree.GetRoot();
+            var annotations = GetAnnotations(root);
+            if (annotations.IsEmpty)
+            {
+                return;
+            }
+            var model = compilation.GetSemanticModel(tree);
+            var expectedTypes = annotations.SelectAsArray(annotation => annotation.Text);
+            var actualTypes = annotations.SelectAsArray(
+                annotation => ((CSharpDataFlowAnalysis)model.AnalyzeDataFlow(annotation.Expression)).TypeAndNullability.ToDisplayString(TypeSymbolWithAnnotations.DebuggerDisplayFormat));
```

Comment:
>((CSharpDataFlowAnalysis)model.AnalyzeDataFlow(annotation.Expression)) [](start = 30, length = 70)

It feels very inefficient to re-analyze the same body over and over again when multiple annotations belong to the same body